### ISSUE TITLE
Pygame-2.0 is more CPU intensive than pygame-1.9.

### ISSUE THREAD

**Issue by ormorph** [pygame/pygame#2455]
I checked pygame-2.0 under **Raspberry Pi 4**, it turned out that the new pygame is very slow. It turns out that pygame-2.0 cannot be used on slow computers. I ran the following code through `pgzero` [this](https://github.com/Wireframe-Magazine/Code-the-Classics). The processor is under heavy load and is very slow. [Pgzero](https://github.com/lordmauve/pgzero).

**Comment by robertpfeiffer** [comment]
What SDL build did you use, where did you get your pygame?

We are tracking raspi hardware support progress in #2301, if you have any concrete tasks to add, I can add them there.

**Comment by ormorph** [comment]
>What SDL build did you use, where did you get your pygame?

Gentoo OS. Installed `media-libs/libsdl2-2.0.14`, `dev-python/pygame-2.0.1`.
If interested, I can make an installation image and upload it to a google drive.

**Comment by ankith26** [comment]
Hmm so you are not using standard raspi OS. I use a pi 3 (latest raspberry pi OS) and have not seen any noticeable performance change between pygame 1.9 and pygame 2.0.1 
I haven’t done any benchmarking or stress testing, though.

The last time I played those code-classics games on my pi, they did seem to run fine, although I agree that they were not as smooth as they would be on my win10 machine.
But it was quite good for the pi anyways

**Comment by ormorph** [comment]
Unfortunately for Pi3 I cannot make an installation image, or rather I can, but that it will work on Pi 3 normally there is no guarantee. I only have Pi 4 available. In Gentoo, you can choose the version of pygame. It is also possible to install the current version from git. The portage system allows you to do this without manual assembly. 
The performance difference is quite large. Comparative tests are not even required, all this is visible by eye.

**Comment by ankith26** [comment]
Maybe it’s a gentoo+rpi specific thing.

**Comment by ormorph** [comment]
>Maybe it’s a gentoo+rpi specific thing.
Also, IIRC, pgzero does not properly support pygame 2.0.0 yet, so I’m not sure how you got that combo working

[message](https://github.com/pygame/pygame/issues/2301#issuecomment-758076032) Judging by the link, this happens not only for me. For pygame-2.0.1 in pgzero, the changes only concern dependencies in setup.py.

[Video](https://gofile.io/d/koRm3y)
There are two files attached in the link above.
**pygame-2.0.1**: `VID_1608102196-16-12-2020.mp4`
**pygame-1.9.6**: `VID_1611050909-19-01-2021.mp4`
The processor in gkrellm loads more pygame-2.0.1.

**Comment by MyreMylar** [comment]
Just a reminder that if this test is doing any sort of alpha blitting with a raspberry pi build of pygame 2 built without the neon extensions it will be slow. 

There is [an issue here](https://github.com/pygame/pygame/issues/1723) relating to trying to get these extensions enabled by default on complicated platforms like the Pi where a single python wheel needs to work on different, incompatible CPU architectures (Pi Zero, Pi 1, Pi 2 Pi 3, Pi 4 ).

In pygame 1/SDL 1 pygame was just using what SDL supplied for blitting in 95% of cases. When pygame 2 upgraded to SDL 2, we discovered its graphics blitter produced substantially different results in some cases  which broke a lot of older games written for pygame 1. Thus a new group of CPU blitting functions were written, just for pygame to cover these cases. To make these CPU blitters fast enough they have to use SIMD extensions (SSE family on Intel/AMD processors and the NEON family on Arm) unfortunately the Pi only has these NEON extensions on it's CPUs in the most popular versions of the Pi. We think there _might_ be a way to build a pygame wheel that works on all forms of Pi, but implementing and testing it is not trivial because we'd need to buy/seek out a lot of different models of Pi and get them all into a state where they could run our test wheel.

Incidentally they have also had different problems with them on the SDL2 side and I believe even SDL2s NEON optimisations may have been disabled recently as well, I'm slightly behind on following where that has ended up.


TLDR; depending on what your benchmark is testing you may need to build both SDL 2 and pygame with specific flags. For pygame it is adding `-enable-arm-neon` when running `python setup.py install`:

```
python setup.py install -enable-arm-neon
```

Assuming you know generally how to build pygame from source.

**Comment by MyreMylar** [comment]
You can also bypass the pygame alpha blitter and use whatever is currently in SDL2 if your use case isn't affected by the bugs & differences in the implementation. To do this see the documentation in the pygame docs on the `PYGAME_BLEND_ALPHA_SDL2` environment variable at the bottom of this page:

https://www.pygame.org/docs/ref/pygame.html

**Comment by ankith26** [comment]
> Hmm so you are not using standard raspi OS. I use a pi 3 (latest raspberry pi OS) and have not seen any noticeable performance change between pygame 1.9 and pygame 2.0.1

Correction. Today I ran some real games on my pi 3 and I saw that there was indeed a noticeable difference in the speed between pygame 1.9.4 and pygame 2.0.1. While I usually ran games with 2.0.1, I did get slowness. But I attributed that to the raspberry pis tiny computing capabilities. And today I realised that pygame 1.9.4 is much faster than pygame 2.0.1 on the pi. Like MyreMylar said, the new alpha blitter is to partially blame, but I suspect that there are more sides to this issue. I will try to do further investigation on this (get the arm neon enabled and some benchmarking and testing with other pygame versions as well)

**Comment by ormorph** [comment]
Later I will try to make a pygame ebuild with `neon` enabled, if everything works, I will add PR to [genpi64-overlay](https://github.com/GenPi64/genpi64-overlay). I will write about the results.
Thanks!

**Comment by ankith26** [comment]
Ok, so I build pygame main branch with arm-neon-enabled and yes, it gives quite some significant speed boost to pygame, while running some games. However, I can’t help but notice that pygame 1.9.4 still performs slightly better when compared to pygame2, even with arm neon enabled. In any case, arm neon optimisations are a boon, and we should look into ways to enable it by default. I will do some further discussion on this on the other issue

**Comment by robertpfeiffer** [comment]
The remaining performance loss may be from pygame.display.update(), which is not as optimised for software rendering as it used to be.

**Comment by robertpfeiffer** [comment]
> Later I will try to make a pygame ebuild with `neon` enabled, if everything works, I will add PR to [genpi64-overlay](https://github.com/GenPi64/genpi64-overlay). I will write about the results.
> Thanks!

That's some really useful information about your problem! Looks like the SDL2 ebuild is a bit out of date 
https://github.com/GenPi64/genpi64-overlay/blob/master/media-libs/libsdl2/libsdl2-2.0.12-r1.ebuild compared to upstream gentoo https://gitweb.gentoo.org/repo/gentoo.git/tree/media-libs/libsdl2/libsdl2-2.0.14-r1.ebuild

SDL2 ships with its own fast blitters, and you may have some luck configuring your SDL2 with 
`--enable-arm-simd` and/or `--enable-arm-neon`, but you have to benchmark them first. It looks like the SDL2 ebuilds have flags for cocoa and 3dnow, but not for ARM.

**Comment by ormorph** [comment]
Checked, unfortunately `neon` is not supported by `aarch64`.  This only works on `arm` 32 bit. Declaring the `PYGAME_BLEND_ALPHA_SDL2=1` variable helped. It is enough to declare a variable with any value.
>Looks like the SDL2 ebuild is a bit out of date

I am using SDL from gentoo upstream, version 2.0.14. After adding the `PYGAME_BLEND_ALPHA_SDL2` variable, the load on the processor disappeared.

**Comment by robertpfeiffer** [comment]
> Checked, unfortunately neon is not supported by aarch64. 

Does --enable-arm-simd work with SDL2? Does it help with PYGAME_BLEND_ALPHA_SDL2=1?

**Comment by ormorph** [comment]
>Does --enable-arm-simd work with SDL2?

This option is not included in the libsdl2 build. 
For sdl2 - `USE="X alsa dbus gles2 haptic joystick opengl pulseaudio sound threads udev video"`

```
./configure --prefix=/usr --build=aarch64-unknown-linux-gnu --host=aarch64-unknown-linux-gnu --mandir=/usr/share/man --infodir=/usr/share/info --datadir=/usr/share --sysconfdir=/etc --localstatedir=/var/lib --docdir=/usr/share/doc/libsdl2-2.0.14 --htmldir=/usr/share/doc/libsdl2-2.0.14/html --with-sysroot=/ --libdir=/usr/lib64 --disable-static --enable-atomic --enable-audio --enable-video --enable-render --enable-events --enable-joystick --enable-haptic --enable-power --enable-filesystem --enable-threads --enable-timers --enable-file --enable-loadso --enable-cpuinfo --enable-assembly --disable-altivec --disable-ssemath --disable-mmx --disable-3dnow --disable-sse --disable-sse2 --disable-oss --enable-alsa --disable-alsa-shared --disable-jack --disable-jack-shared --disable-esd --enable-pulseaudio --disable-pulseaudio-shared --disable-arts --disable-libsamplerate --disable-nas --disable-nas-shared --disable-sndio --disable-sndio-shared --enable-diskaudio --enable-dummyaudio --disable-video-wayland --disable-wayland-shared --enable-video-rpi --enable-video-x11 --disable-x11-shared --enable-video-x11-xcursor --enable-video-x11-xdbe --disable-video-x11-xinerama --enable-video-x11-xinput --enable-video-x11-xrandr --disable-video-x11-scrnsaver --enable-video-x11-xshape --enable-video-x11-vm --disable-video-cocoa --disable-video-directfb --disable-fusionsound --disable-fusionsound-shared --disable-video-kmsdrm --disable-kmsdrm-shared --enable-video-dummy --enable-video-opengl --disable-video-opengles1 --enable-video-opengles2 --disable-video-vulkan --enable-libudev --enable-dbus --disable-fcitx --disable-ibus --disable-directx --disable-rpath --disable-render-d3d --with-x
```

**Comment by robertpfeiffer** [comment]
> This option is not included in the libsdl2 build.

Yes. I read the ebuild files I linked to. I am asking again: If you enable these options (e.g. by adding them to your ebuild file) does the SDL2 blitting performance (enabled with PYGAME_BLEND_ALPHA_SDL2=1) improve?

also @MyreMylar this might become relevant for M1 mac support

**Comment by robertpfeiffer** [comment]
@ormorph can you confirm that the -enable-arm-neon optimisations in PyGame and --enable-arm-neon in SDL2 do not work on aarch64 hardware, or is this just a bug in the Gentoo system's implementation of the `neon` USE flag?

If that is the case, please report a bug with Gentoo.

Please report a bug with Gentoo anyway. The SDL2 ebuild is missing USE flags that should really speed up PYGAME_BLEND_ALPHA_SDL2=1 on the systems where the Neon instruction set is available, and even if you could enable `neon`, it the ebuild does not check the flag to actually pass --enable-arm-neon to the configure script.

**Comment by ormorph** [comment]
>@ormorph can you confirm that the -enable-arm-neon optimisations in PyGame and --enable-arm-neon in SDL2 do not work on aarch64 hardware, or is this just a bug in the Gentoo system's implementation of the neon USE flag?

gcc aarch64 doesn't support build flag `-mfpu=neon`. There is nothing to confirm here. This flag only works on arm 32 bits. In Gentoo, the neon flag is masked. 
After declaring the `PYGAME_BLEND_ALPHA_SDL2=1` variable, the processor load is about the same as on the Raspberry Pi OS. There is no point in adding the `--enable-arm-simd` option yet. The processor is not even loaded.
While busy adding [RPi.GPIO](https://sourceforge.net/projects/raspberry-gpio-python) to genpi64, `gpiozero` is needed. I'll post the ebuild later.

**Comment by ankith26** [comment]
> gcc aarch64 doesn't support build flag `-mfpu=neon`. There is nothing to confirm here. This flag only works on arm 32 bits. In Gentoo, the neon flag is masked.

True. I guess the correct flag for those optimisations on an aarch64 would be.
`-march=armv8-a+fp+simd`
I’m not a 100 percent sure, but I got this flag from the sse2neon library readme (https://github.com/DLTcollab/sse2neon).
Pygame uses this library for arm-optimizing the alpha-blit stuff

We might need to make pygame source changes to account this in

**Comment by robertpfeiffer** [comment]
> gcc aarch64 doesn't support build flag -mfpu=neon

Ah, now I understand. That's not what I meant though. I know armv8 still supports neon in some form, although if you do not use intrinsics (which will still work, I looked it up in the official armv8 docs), you may have to make some changes in your assembly source.

I'll just ask the SDL people to document `--enable-arm-simd` and `--enable-arm-neon` better.  I know SDL2 has some ARM SIMD assembly code in `src/video/arm/pixman-arm-simd-asm.S`, but I don't know how portable that is. That is not the same as `-mfpu=neon`.

As for pygame, I thought we are already using sse2neon intrinsics, so why can't we enable them and let the compiler figure out whether to generate SSE4.2, i386, armv6, or neon?

> True. I guess the correct flag for those optimisations on an aarch64 would be. `-march=armv8-a+fp+simd`
Somewhat bafflingly, `-march=native` does not work on most architectures, so really, you need to detect what cpu you are running, and then pass the right `-march` parameter on arm.

Using -march=native by default would be a bad idea for building wheels or distributing games with pyinstaller anyway.

**Comment by ormorph** [comment]
Compared the performance of pygame versions. It turned out that version 2.0.1 of the declared variable `PYGAME_BLEND_ALPHA_SDL2=1`, about half the load on the processor than version 1.9.x.

**Comment by robertpfeiffer** [comment]
So does anybody any idea what happened here? Looks like pygame was compiled without sse2neon intrinsics, or is neon enabled by default with the aarch64 GCC backend? I assume SDL2 was configured without the ARM+neon specific assembly code, but the aarch64 gcc backend still generated optimised enough code?

**Comment by ormorph** [comment]
Quote from the [link]:(https://github.com/GenPi64/gentoo-on-rpi-64bit)

>Note: the CFLAGS used for the image build is -march=armv8-a+crc -mtune=cortex-a72 -O2 -pipe. You can of course re-build selective components with more aggressive flags yourself, should you choose. As the SIMD FPU features are standard in ARMv8, there is no need for -mfpu=neon mfloat-abi=hard etc., as you would have had on e.g. the 32-bit ARMv7a architecture. Note that AArch64 NEON also has a full IEEE 754-compliant mode (including handling denormalized numbers etc.), there is also no need for -ffast-math flag to fully exploit the FPU either (again, unlike earlier ARM systems). Please refer to the official Programmer’s Guide for ARMv8-A for more details. Neither the Pi3 nor Pi4 is shipped with the ARMv8 hardware crypto extensions enabled, unfortunately.

**Comment by robertpfeiffer** [comment]
Yes. This is why I was so confused. If neon is enabled by default, PyGgame should be built with sse2neon intrinsics (which is an implementation of SSE intrinsics with neon instructions). Maybe these cflags were just used by emerge, and not by pip and setup.py.

On the other hand, SDL2 was configured (in the ebuild) without `--enable-assembly` and without `--enable-arm-simd`. So the assembly code in `src/video/arm/pixman-arm-simd-asm.S` was not compiled into SDL2. Even if you pass `-march=armv8-a` or `-mcpu=native` to gcc, aggressive vectorisation and loop unrolling are only enabled at -O3, so the gcc should not emit neon/SIMD instructions for the SDL2 blitting code. I do not know if the armv6-specific and armv7-specific handwritten assembly code would even work on or run faster on 64-bit armv8 than the optimised code emitted by gcc.

That leaves a puzzle: You compiled SDL2 without vectorisation and without the fast assembly code path, but Pygame with neon intrinsics on armv8, yet PYGAME_BLEND_ALPHA_SDL2=1 has a huge performance gain.

**Comment by ormorph** [comment]
>That leaves a puzzle: You compiled SDL2 without vectorisation and without the fast assembly code path, but Pygame with neon intrinsics on armv8, yet PYGAME_BLEND_ALPHA_SDL2=1 has a huge performance gain.

Hmm, a complete list of build parameters can be viewed [here](https://github.com/GenPi64/genpi64-overlay/blob/master/profiles/targets/genpi64/make.defaults). Vectorization has been enabled...
These options are enabled when one of the genpi64 profiles is selected.

**Comment by MyreMylar** [comment]
The answer to the puzzle is that `-enable-arm-neon` in pygame doesn't just add the cpu flag `-mfpu=neon`, it also defs in all the NEON code via the `PG_ENABLE_ARM_NEON` flag.

Probably we need to change line 142 of setup.py:

`    cflags += '-mfpu=neon'`

with a per cpu if like @ankith26 suggests above so it only adds this flag on arm7, not arm6 (which doesn't have them) or arm8 (which apparently has Neon as always on).

**Comment by robertpfeiffer** [comment]
CPU dispatch in PyGame is subtly broken both on x86 and ARM.

**Comment by robertpfeiffer** [comment]
Still working on getting CPU dispatch to run. We probably need to decide which compilers we support fully. I assume we need at least GCC, MSVC, and clang. Other compilers could be supported in some limited capacity. That means at least making it possible to build, even dispatch is not working.

**Comment by ormorph** [comment]
See there is no optimization for aarch64 in alphablit.c. The declaration of the `PG_ENABLE_ARM_NEON` macro will only work on 32 bit arm.  It seems to use variables specific to a 32-bit processor.  It turns out that optimization for `aarch64` is not yet provided.
